gravatar-quickeditor/src/main/java/com/gravatar/quickeditor/ui/alttext/AltTextViewModel.kt

```diff
@@ -0,0 +1,101 @@
+import androidx.lifecycle.ViewModelProvider
+import androidx.lifecycle.viewModelScope
+import androidx.lifecycle.viewmodel.CreationExtras
+import com.gravatar.quickeditor.QuickEditorContainer
+import com.gravatar.quickeditor.data.repository.AvatarRepository
+import com.gravatar.services.GravatarResult
+import com.gravatar.types.Email
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.receiveAsFlow
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+
+internal class AltTextViewModel(
+    private val email: String,
+    private val avatarId: String,
+    private val altText: String,
+    avatarUrl: String,
+    private val avatarRepository: AvatarRepository,
+) : ViewModel() {
+    private val _uiState =
+        MutableStateFlow(
+            AltTextUiState(avatarUrl = avatarUrl, isSaveButtonEnabled = false, isUpdating = false, altText = altText),
+        )
+    val uiState: StateFlow<AltTextUiState> = _uiState.asStateFlow()
+    private val _actions = Channel<AltTextAction>(Channel.BUFFERED)
+    val actions = _actions.receiveAsFlow()
+
+    fun onEvent(event: AltTextEvent) {
+        when (event) {
+            is AltTextEvent.AvatarAltTextSaveTapped -> updateAltText()
+            is AltTextEvent.AvatarAltTextChange -> updateUiStateWithNewAltText(event.newAltText)
+        }
+    }
+
+    private fun updateAltText() {
+        viewModelScope.launch {
+            _uiState.update { currentState ->
+                currentState.copy(
+                    isSaveButtonEnabled = false,
+                    isUpdating = true,
+                )
+            }
+            when (avatarRepository.updateAvatar(Email(email), avatarId, altText = uiState.value.altText)) {
+                is GravatarResult.Success -> {
+                    _uiState.update { currentState ->
+                        currentState.copy(
+                            isSaveButtonEnabled = true,
+                            isUpdating = false,
+                        )
+                    }
+                    _actions.send(AltTextAction.AltTextUpdated)
+                }
+
+                is GravatarResult.Failure -> {
+                    _uiState.update { currentState ->
+                        currentState.copy(
+                            isSaveButtonEnabled = true,
+                            isUpdating = false,
+                        )
+                    }
+                    _actions.send(AltTextAction.AltTextUpdateFailed)
+                }
+            }
+        }
+    }
+
+    private fun updateUiStateWithNewAltText(newAltText: String) {
+        viewModelScope.launch {
+            _uiState.update { currentState ->
+                currentState.copy(
+                    altText = newAltText,
+                    isSaveButtonEnabled = newAltText != altText,
```

Comment:
When the `newAltTest` is empty the API call returns 200, but it doesn't update it. Looks like it quietly fails. Should we disable the save button in such cases?

Reply:
🤔 Nice catch! I haven't noticed that.

The web is applying a "default" `AltText` when trying to save it as empty (Example: `Héctor Abraham's avatar`). The endpoint we are using is not working the same way. I'll check with the backend folks. If we want to set the default `AltText`, we need the backend to return the avatar object with that data.

Reply:
It looks like this will be fixed in the backend and should be transparent for us (as we use the returned avatar object) so I think we can merge the PR.

Reply:
Sounds good!